File: 102_BinaryTreeLevelOrderTraversal.cpp
/** Binary Tree Level Order Traversal
Given a binary tree, return the level order traversal of its nodes' values. (ie, from left to right, level by level).

For example:
Given binary tree {3,9,20,#,#,15,7},
    3
   / \
  9  20
    /  \
   15   7
return its level order traversal as:
[
  [3],
  [9,20],
  [15,7]
]
 */
//算法思想，二叉树的按层遍历，我通过递归的方式实现如下
class Solution {
public:

    void level(vector<TreeNode *> presentLevel, vector<vector<int>> &result){
        vector<int> tmp;
        vector<TreeNode *> tmpTreeNode;
        
        int n = presentLevel.size();
        if(n <= 0) return;
        for(int i=0; i<n; i++){
            tmp.push_back(presentLevel[i]->val);
            if(presentLevel[i]->left != NULL){
                tmpTreeNode.push_back(presentLevel[i]->left);
            }
            if(presentLevel[i]->right != NULL){
                tmpTreeNode.push_back(presentLevel[i]->right);
            }
        }
        result.push_back(tmp);
        level(tmpTreeNode, result);
    }
    vector<vector<int>> levelOrder(TreeNode* root) {
        
        vector<vector<int>> result;
        vector<TreeNode*> firstLevel;
        
        if(root != NULL){
            firstLevel.push_back(root);
            level(firstLevel, result);
        }
        
        return result;
        
    }
};

//利用递归的思想，通过栈来实现，非递归的方式，用另外一个队列来保存孩子节点层
class Solution {
public:

    vector<vector<int>> levelOrder(TreeNode* root) {
        
        vector<vector<int>> result;
        vector<TreeNode*> level;
        
        if(root != NULL){
            level.push_back(root);
        }
        while(level.size() > 0){
            vector<int> tmp;
            vector<TreeNode*> nextLevel;
            for(int i=0; i<level.size(); i++){
                tmp.push_back(level[i]->val);
                if(level[i]->left != NULL){
                    nextLevel.push_back(level[i]->left);
                }
                if(level[i]->right != NULL){
                    nextLevel.push_back(level[i]->right);
                }
            }
            result.push_back(tmp);
            level = nextLevel;
        }
        
        return result;
        
    }
};

File: 102-binary-tree-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        if (root == null) {
            return Collections.emptyList();
        }
        List<TreeNode> list = new ArrayList<TreeNode>();
        list.add(root);
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        helper(list, result);
        return result;
    }

    private void helper(List<TreeNode> list, List<List<Integer>> result) {
        if (list == null || list.size() == 0) {
            return;
        }
        List<TreeNode> children = new ArrayList<TreeNode>();
        List<Integer> resultList = new ArrayList<Integer>();
        for (TreeNode node : list){
            resultList.add(node.val);
            if (node.left != null) {
                children.add(node.left);
            }
            if (node.right != null){
                children.add(node.right);
            }
        }
        result.add(resultList);
        helper(children, result);
    }
}

class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        if (root == null){
            return Collections.emptyList();
        }
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        helper(result, root, 1);
        return result;
    }

    private void helper(List<List<Integer>> result, TreeNode root, int level) {
        if (root == null) {
            return;
        }
        if (result.size() < level) {
            List<Integer> list = new ArrayList<Integer>();
            result.add(list);
        }
        List<Integer> nums = result.get(level-1);
        nums.add(root.val);
        helper(result, root.left, level + 1);
        helper(result, root.right, level + 1);
    }
}
